File: 785.IsGraphBipartite.py
class Solution(object):
    def isBipartite(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: bool
        """
        color = [0] * len(graph)
        for i in range(len(graph)):
            if color[i] == 0 and not self.dfs(graph, color, i):
                return False

    def dfs(self, graph, color, i):
        color[i] = 1
        q = [i]
        while q:
            i = q.pop(0)
            for neighbor in graph[i]:
                if color[neighbor] == 0:
                    q.append(neighbor)
                    color[neighbor] = -color[i]
                elif color[neighbor] == color[i]:
                    return False
        return True

File: Algorithms/cpp/785.IsGraphBipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int N = graph.size();
        int color[N];
        for (int i=0; i < N; i++) {
            color[i] = 0;
        }
        for (int i=0; i < N; i++) {
            if (color[i] == 0 && !dfs(graph, color, i)) {
                return false;
            }
        }
        return true;
    }

private: 
    bool dfs(vector<vector<int>>& graph, int color[], int j) {
        color[j] = 1;
        queue<int> q;
        q.push(j);
        while(!q.empty()) {
            int i = q.front();
            q.pop();
            for (int k=0; k < graph[i].size(); k++) {
                int neighbor = graph[i][k];
                if (color[neighbor] == 0) {
                    q.push(neighbor);
                    color[neighbor] = -color[i];
                }
                else if (color[neighbor] == color[i]) {
                    return false;
                }
            }
        }
        return true;
    }
};
